src/bitcoin_monitor/bitcoin/tracepoints/net.py

```diff
@@ -0,0 +1,306 @@
+"""Log ip accounting data for bitcoind."""
+
+import asyncio
+import csv
+import datetime
+import logging as log
+import subprocess
+import time
+from dataclasses import dataclass
+from pathlib import Path
+from typing import ClassVar
+
+import psutil
+from bcc import BPF, USDT
+
+
+class Message:
+    """A P2P network message."""
+
+    msg_type = ""
+    size = 0
+    data = bytes()
+    inbound = False
+
+    def __init__(self, msg_type, size, inbound):
+        self.msg_type = msg_type
+        self.size = size
+        self.inbound = inbound
+
+
+class Peer:
+    """A P2P network peer."""
+
+    id = 0
+    address = ""
+    connection_type = ""
+    last_messages = list()
+
+    total_inbound_msgs = 0
+    total_inbound_bytes = 0
+    total_outbound_msgs = 0
+    total_outbound_bytes = 0
+
+    def __init__(self, id, address, connection_type):
+        self.id = id
+        self.address = address
+        self.connection_type = connection_type
+        self.last_messages = list()
+
+    def add_message(self, message):
+        self.last_messages.append(message)
+        if len(self.last_messages) > 25:
+            self.last_messages.pop(0)
+        if message.inbound:
+            self.total_inbound_bytes += message.size
+            self.total_inbound_msgs += 1
+        else:
+            self.total_outbound_bytes += message.size
+            self.total_outbound_msgs += 1
+
+
+PROGRAM = """
+#include <uapi/linux/ptrace.h>
+
+// Tor v3 addresses are 62 chars + 6 chars for the port (':12345').
+// I2P addresses are 60 chars + 6 chars for the port (':12345').
+#define MAX_PEER_ADDR_LENGTH 62 + 6
+#define MAX_PEER_CONN_TYPE_LENGTH 20
+#define MAX_MSG_TYPE_LENGTH 20
+
+struct p2p_message
+{
+    u64     peer_id;
+    char    peer_addr[MAX_PEER_ADDR_LENGTH];
+    char    peer_conn_type[MAX_PEER_CONN_TYPE_LENGTH];
+    char    msg_type[MAX_MSG_TYPE_LENGTH];
+    u64     msg_size;
+};
+
+
+// Two BPF perf buffers for pushing data (here P2P messages) to user space.
+BPF_PERF_OUTPUT(inbound_messages);
+BPF_PERF_OUTPUT(outbound_messages);
+
+int trace_inbound_message(struct pt_regs *ctx) {
+    struct p2p_message msg = {};
+
+    bpf_usdt_readarg(1, ctx, &msg.peer_id);
+    bpf_usdt_readarg_p(2, ctx, &msg.peer_addr, MAX_PEER_ADDR_LENGTH);
+    bpf_usdt_readarg_p(3, ctx, &msg.peer_conn_type, MAX_PEER_CONN_TYPE_LENGTH);
+    bpf_usdt_readarg_p(4, ctx, &msg.msg_type, MAX_MSG_TYPE_LENGTH);
+    bpf_usdt_readarg(5, ctx, &msg.msg_size);
```

Comment:
I think it makes sense to avoid using `bpf_usdt_readarg_p` going forward. See https://github.com/bitcoin/bitcoin/issues/27380#issuecomment-2500962838 and the following comment. I plan to open a PR that replaces the bitcoin/bitcoin usage. You can pick these changes up then.